File: src/main/java/wraith/fabricaeexnihilo/api/recipes/barrel/FluidOnTopRecipe.java
package wraith.fabricaeexnihilo.api.recipes.barrel;

import alexiil.mc.lib.attributes.fluid.volume.FluidVolume;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.FluidBlock;
import wraith.fabricaeexnihilo.api.crafting.FluidIngredient;
import wraith.fabricaeexnihilo.modules.barrels.modes.BarrelMode;

public record FluidOnTopRecipe(FluidIngredient inBarrel, FluidIngredient onTop, BarrelMode result) {

    public boolean test(FluidVolume contents, FluidBlock block) {
        return inBarrel.test(contents) && onTop.test(block);
    }

    public boolean test(FluidVolume contents, Block block) {
        return inBarrel.test(contents) && onTop.test(block);
    }

    public boolean test(FluidVolume contents, BlockState block) {
        return inBarrel.test(contents) && onTop.test(block);
    }

}

File: src/main/kotlin/exnihilofabrico/api/recipes/barrel/FluidOnTopRecipe.kt
package exnihilofabrico.api.recipes.barrel

import alexiil.mc.lib.attributes.fluid.volume.FluidVolume
import exnihilofabrico.api.crafting.FluidIngredient
import exnihilofabrico.modules.barrels.modes.BarrelMode
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.FluidBlock

data class FluidOnTopRecipe(val inBarrel: FluidIngredient, val onTop: FluidIngredient, val result: BarrelMode) {
    fun test(contents: FluidVolume, block: FluidBlock) = inBarrel.test(contents) && onTop.test(block)
    fun test(contents: FluidVolume, block: Block) = inBarrel.test(contents) && onTop.test(block)
    fun test(contents: FluidVolume, block: BlockState) = inBarrel.test(contents) && onTop.test(block)
}